x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/AbstractProfilingPersistenceManager.java

```diff
@@ -0,0 +1,226 @@
+            return;
+        }
+        // wait for the cluster state to be recovered
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            return;
+        }
+
+        // If this node is not a master node, exit.
+        if (event.state().nodes().isLocalNodeElectedMaster() == false) {
+            return;
+        }
+
+        if (event.state().nodes().getMaxNodeVersion().after(event.state().nodes().getSmallestNonClientNodeVersion())) {
+            logger.debug("Skipping up-to-date check as cluster has mixed versions");
+            return;
+        }
+
+        if (isAllResourcesCreated(event) == false) {
+            logger.trace("Skipping index creation; not all required resources are present yet");
+            return;
+        }
+
+        if (inProgress.compareAndSet(false, true)) {
+            // Only release the lock once all upgrade attempts have succeeded or failed.
+            try (var refs = new RefCountingRunnable(() -> inProgress.set(false))) {
+                ClusterState clusterState = event.state();
+                for (T index : getManagedIndices()) {
+                    Status status = getStatus(clusterState, index);
+                    if (status.actionable) {
+                        onStatus(clusterState, status, index, ActionListener.releasing(refs.acquire()));
+                    }
+                }
+            }
+        } else {
+            logger.trace("Skipping index creation as changes are already in progress");
+        }
+    }
+
+    protected boolean isAllResourcesCreated(ClusterChangedEvent event) {
+        return ProfilingIndexTemplateRegistry.isAllResourcesCreated(event.state());
+    }
+
+    /**
+     * Extracts the appropriate index metadata for a given index from the cluster state.
+     *
+     * @param state Current cluster state. Never <code>null</code>.
+     * @param index An index for which to retrieve index metadata. Never <code>null</code>.
+     * @return The corresponding index metadata or <code>null</code> if there are none.
+     */
+    protected abstract IndexMetadata indexMetadata(ClusterState state, T index);
+
+    /**
+     * @return An iterable of all indices that are managed by this instance.
+     */
+    protected abstract Iterable<T> getManagedIndices();
+
+    /**
+     * Handler that takes appropriate action for a certain index status.
+     *
+     * @param clusterState The current cluster state. Never <code>null</code>.
+     * @param status Status of the current index.
+     * @param index The current index.
+     * @param listener Listener to be called on completion / errors.
+     */
+    protected abstract void onStatus(ClusterState clusterState, Status status, T index, ActionListener<? super ActionResponse> listener);
+
+    private Status getStatus(ClusterState state, T index) {
+        IndexMetadata metadata = indexMetadata(state, index);
+        if (metadata != null) {
```

Comment:
If you return early here and for other return conditions, a lot of indenting can be removed. IMO, that would make the code more readable.

Reply:
Good point. Done in https://github.com/elastic/elasticsearch/pull/97380/commits/5d88466207834582b22c8227aacc8dbb88413ac7.